namespace/http_resolver.go

```diff
@@ -0,0 +1,423 @@
+				} else {
+					parsedScope = scp
+				}
+			default:
+				return nil, fmt.Errorf("unexpected html element %q", string(tagName))
+			}
+			readingMetaTags = true
+		case html.EndTagToken:
+			switch tagName, _ := z.TagName(); string(tagName) {
+			case "head":
+				break ParsingLoop
+			case "meta":
+			default:
+				return nil, fmt.Errorf("unexpected tag %q", tagName)
+			}
+		default:
+			continue ParsingLoop
+		}
+	}
+	if !readingMetaTags || len(entries.entries) == 0 {
+		return nil, fmt.Errorf("no entries found")
+	}
+	if parsedScope == "" {
+		parsedScope, err = parseScope(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot use given name %q as scope: %v", name, err)
+		}
+	}
+	for i := range entries.entries {
+		entries.entries[i].scope = parsedScope
+	}
+	return entries, nil
+}
+
+type httpResolver struct {
+	config *HTTPResolverConfig
+	cache  map[string]cacheEntry
+}
+
+/* Create base HTTP resolver.
+ *
+ * It uses discovery process to fetch scope entries. Multiple discovery
+ * endpoints may be queried during single `Resolve()` call depending on scope's
+ * namespace extensions and given callback.
+ *
+ * It caches fetched entries for further lookups.
+ */
+func NewHTTPResolver(config *HTTPResolverConfig) Resolver {
+	if config == nil {
+		config = &HTTPResolverConfig{}
+	}
+	if config.Client == nil {
+		config.Client = &http.Client{}
+	}
+	if config.ResolverFactory == nil {
+		config.ResolverFactory = func(entries *Entries) Resolver {
+			return NewSimpleResolver(entries, true)
+		}
+	}
+	if config.NSResolveCallback == nil {
+		config.NSResolveCallback = func(name string, namespace scope) NSResolveActionEnum {
+			if !namespace.Contains(name) {
+				logrus.Debugf("Ignoring extension namespace %q which isn't an ancestor of %q", namespace, name)
+				return NSResolveActionIgnore
+			}
+			return NSResolveActionRecurse
+		}
+	}
+	return &httpResolver{
+		config: config,
+		cache:  make(map[string]cacheEntry),
+	}
+}
+
+func (hr *httpResolver) nameToURL(name string) string {
+	return "https://" + name + "?docker-discovery=1"
+}
+
+func (hr *httpResolver) resolveEntries(es *Entries, visited map[string]struct{}, name string) error {
+	cached, exists := hr.cache[name]
+	if exists && !cached.created.Add(hr.config.ExpireAfter).Before(time.Now()) {
+		entries, err := hr.config.ResolverFactory(cached.entries).Resolve(name)
+		if err != nil {
+			return err
+		}
+		entries, err = es.Join(entries)
+		if err != nil {
+			return err
+		}
+		*es = *entries
+		return nil
+	}
+	if exists {
+		delete(hr.cache, name)
+	}
+	resp, err := hr.config.Client.Get(hr.nameToURL(name))
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("discovery endpoint %q replied with: %s", name, resp.Status)
+	}
+	defer resp.Body.Close()
+	// TODO: check content type
+
+	entries, err := parseHTMLHead(resp.Body, name)
+	if err != nil {
+		return err
+	}
+
+	// handle scope extensions
```

Comment:
The recursion shouldn't be necessary here.  Since it is common functionality it should be done in either the MultiResolver or another implementation which is able to recurse.